File: src/editor/gui/display/CardImagePanel.java
package editor.gui.display;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

import javax.imageio.ImageIO;
import javax.swing.JPanel;
import javax.swing.JTextPane;
import javax.swing.SwingUtilities;
import javax.swing.SwingWorker;
import javax.swing.UIManager;
import javax.swing.text.Style;
import javax.swing.text.StyleConstants;
import javax.swing.text.StyledDocument;

import editor.database.card.Card;
import editor.database.card.CardLayout;
import editor.gui.MainFrame;
import editor.gui.SettingsDialog;

/**
 * This class represents a panel that shows the images associated with a card if they
 * can be found, or a card-shaped rectangle with its oracle text and a warning if
 * they cannot.
 *
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class CardImagePanel extends JPanel
{
    /**
     * Aspect ratio of a Magic: The Gathering card.
     */
    public static final double ASPECT_RATIO = 63.0/88.0;

    /**
     * This class represents a request by a CardImagePanel to download the image(s)
     * of a Card.
     */
    private static class DownloadRequest
    {
        /**
         * CardImagePanel that needs to download a card.
         */
        public final CardImagePanel source;
        /**
         * Card that needs to be downloaded.
         */
        public final Card card;

        /**
         * Create a new DownloadRequest.
         * 
         * @param panel CardImagePanel making the request
         * @param c Card whose images need to be downloaded
         */
        public DownloadRequest(CardImagePanel panel, Card c)
        {
            source = panel;
            card = c;
        }
    }

    /**
     * This class represents a worker that downloads a card image for its parent CardImagePanel
     * from Gatherer.
     */
    private static class ImageDownloadWorker extends SwingWorker<Void, DownloadRequest>
    {
        /**
         * Queue of cards whose images still need to be downloaded.
         */
        private BlockingQueue<DownloadRequest> toDownload;

        /**
         * Create a new ImageDownloadWorker.
         */
        public ImageDownloadWorker()
        {
            super();
            toDownload = new LinkedBlockingQueue<>();
        }

        /**
         * Create a request to download the image(s) of a card to display on the given
         * CardImagePanel.
         */
        public void downloadCard(CardImagePanel source, Card card)
        {
            try
            {
                toDownload.put(new DownloadRequest(source, card));
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
        }

        @Override
        protected Void doInBackground() throws Exception
        {
            while (true)
            {
                DownloadRequest req = toDownload.take();
                for (long multiverseid : req.card.multiverseid())
                {
                    File img = Paths.get(SettingsDialog.getAsString(SettingsDialog.CARD_SCANS), multiverseid + ".jpg").toFile();
                    if (!img.exists())
                    {
                        URL site = new URL(String.join("/", "http://gatherer.wizards.com", "Handlers", "Image.ashx?multiverseid=" + multiverseid + "&type=card"));

                        img.getParentFile().mkdirs();
                        // TODO: Add a timeout here
                        try (BufferedInputStream in = new BufferedInputStream(site.openStream()))
                        {
                            try (BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(img)))
                            {
                                byte[] data = new byte[1024];
                                int x;
                                while ((x = in.read(data)) > 0)
                                    out.write(data, 0, x);
                            }
                        }
                        catch (Exception e)
                        {
                            System.err.println("Error downloading " + multiverseid + ".jpg: " + e.getMessage());
                        }
                    }
                }
                publish(req);
            }
        }

        @Override
        protected void process(List<DownloadRequest> chunks)
        {
            for (DownloadRequest req: chunks)
                if (req.source.card == req.card)
                    req.source.loadImages();
        }
    }

    /**
     * Global ImageDownloadWorker to manage downloading cards images.
     */
    private static ImageDownloadWorker downloader = new ImageDownloadWorker();
    static
    {
        downloader.execute();
    }

    /**
     * This class represents a listener that listens for clicks on a CardImagePanel.
     */
    private class FaceListener extends MouseAdapter
    {
        /**
         * When the mouse is clicked, flip to the next face.
         */
        public void mousePressed(MouseEvent e)
        {
            if (SwingUtilities.isLeftMouseButton(e))
            {
                switch (card.layout())
                {
                case SPLIT:
                    face = 0;
                    break;
                default:
                    face = (face + 1)%card.faces();
                    break;
                }
                getParent().revalidate();
                repaint();
            }
        }
    }

    /**
     * Card this CardImagePanel should display.
     */
    private Card card;
    /**
     * List of images to draw for the card.
     */
    private List<BufferedImage> faceImages;
    /**
     * Image of the card this CardImagePanel should display.
     */
    private BufferedImage image;
    /**
     * Face of the card to display.
     */
    private int face;

    /**
     * Create a new CardImagePanel displaying nothing.
     */
    public CardImagePanel()
    {
        this(null);
    }

    /**
     * Create a new CardImagePanel displaying the specified card.
     *
     * @param c card to display
     */
    public CardImagePanel(Card c)
    {
        super(null);
        image = null;
        faceImages = new ArrayList<>();
        face = 0;
        setCard(c);
        addMouseListener(new FaceListener());
    }

    /**
     * {@inheritDoc}
     * The preferred size is the largest rectangle that fits the image this CardImagePanel is trying
     * to draw that fits within the parent container.
     */
    @Override
    public Dimension getPreferredSize()
    {
        if (getParent() == null)
            return super.getPreferredSize();
        else if (image == null)
            return super.getPreferredSize();
        else
        {
            double aspect = (double)image.getWidth()/(double)image.getHeight();
            return new Dimension((int)(getParent().getHeight()*aspect), getParent().getHeight());
        }
    }

    /**
     * Once the images have been downloaded, try to load them.  If they don't exist,
     * create a rectangle with Oracle text instead.
     */
    private synchronized void loadImages()
    {
        if (card != null)
        {
            faceImages.clear();
            for (long i : card.multiverseid())
            {
                BufferedImage img = null;
                try
                {
                    if (i > 0)
                    {
                        File imageFile = Paths.get(SettingsDialog.getAsString(SettingsDialog.CARD_SCANS), i + ".jpg").toFile();
                        if (imageFile.exists())
                            img = ImageIO.read(imageFile);
                    }
                }
                catch (IOException e)
                {}
                finally
                {
                    faceImages.add(img);
                }
            }
            if (getParent() != null)
            {
                SwingUtilities.invokeLater(() -> {
                    getParent().revalidate();
                    repaint();
                });
            }
        }
    }

    /**
     * {@inheritDoc}
     * The panel will basically just be the image generated in {@link CardImagePanel#setCard(Card)}
     * scaled to fit the container.
     */
    @Override
    protected void paintComponent(Graphics g)
    {
        super.paintComponent(g);
        if (image != null)
        {
            Graphics2D g2 = (Graphics2D)g;
            g2.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BICUBIC);
            double aspectRatio = (double)image.getWidth() / (double)image.getHeight();
            int width = (int)(getHeight() * aspectRatio);
            int height = getHeight();
            if (width > getWidth())
            {
                width = getWidth();
                height = (int)(width / aspectRatio);
            }
            g2.drawImage(image, (getWidth() - width) / 2, (getHeight() - height) / 2, width, height, null);
        }
    }

    /**
     * Set the bounding box of this CardImagePanel.  This will cause it to refresh its image
     * to fit inside the new bounding box.
     */
    @Override
    public void setBounds(int x, int y, int width, int height)
    {
        super.setBounds(x, y, width, height);
        if (card == null || height == 0 || width == 0)
            image = null;
        else
        {
            int h = 0, w = 0;
            if (faceImages.size() < face || faceImages.get(face) == null)
            {
                h = height;
                w = (int)(h*ASPECT_RATIO);
            }
            else
            {
                h = faceImages.get(face).getHeight();
                w = faceImages.get(face).getWidth();
            }
            image = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
            Graphics g = image.createGraphics();
            if (faceImages.size() < face || faceImages.get(face) == null)
            {
                int faceWidth = (int)(h*ASPECT_RATIO);

                JTextPane missingCardPane = new JTextPane();
                StyledDocument document = (StyledDocument)missingCardPane.getDocument();
                Style textStyle = document.addStyle("text", null);
                StyleConstants.setFontFamily(textStyle, UIManager.getFont("Label.font").getFamily());
                StyleConstants.setFontSize(textStyle, MainFrame.TEXT_SIZE);
                Style reminderStyle = document.addStyle("reminder", textStyle);
                StyleConstants.setItalic(reminderStyle, true);
                card.formatDocument(document, false, face);
                missingCardPane.setSize(new Dimension(faceWidth - 4, h - 4));

                BufferedImage img = new BufferedImage(faceWidth, h, BufferedImage.TYPE_INT_ARGB);
                missingCardPane.paint(img.getGraphics());
                g.drawImage(img, 2, 2, null);
                g.setColor(Color.BLACK);
                g.drawRect(0, 0, faceWidth - 1, h - 1);
            }
            else
            {
                if (card.layout() == CardLayout.FLIP && face%2 == 1)
                    g.drawImage(faceImages.get(face), faceImages.get(0).getWidth(), faceImages.get(0).getHeight(), -faceImages.get(0).getWidth(), -faceImages.get(0).getHeight(), null);
                else
                    g.drawImage(faceImages.get(face), 0, 0, null);
            }
        }
    }

    /**
     * Set the card to display.  If its image is missing, try to download it.
     *
     * @param c card to display
     */
    public void setCard(Card c)
    {
        if ((card = c) != null)
        {
            face = 0;
            downloader.downloadCard(this, card);
        }
    }
}

File: src/main/scala/editor/gui/display/CardImagePanel.scala
package editor.gui.display

import editor.database.card.Card
import editor.database.card.CardLayout
import editor.database.symbol.FunctionalSymbol
import editor.gui.generic.ComponentUtils
import editor.gui.settings.SettingsDialog

import java.awt.Color
import java.awt.Dimension
import java.awt.FlowLayout
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.RenderingHints
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.geom.AffineTransform
import java.awt.image.AffineTransformOp
import java.awt.image.BufferedImage
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.URL
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import javax.imageio.ImageIO
import javax.swing.BorderFactory
import javax.swing.Box
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JProgressBar
import javax.swing.JTextPane
import javax.swing.SwingUtilities
import javax.swing.SwingWorker
import javax.swing.UIManager
import javax.swing.text.StyleConstants
import javax.swing.text.StyledDocument
import scala.jdk.CollectionConverters._
import scala.util.Using

/**
 * Structure containing information about card images and where to get them and downloads missing card images when they are
 * set to be displayed in a [[CardImagePanel]].
 * 
 * @author Alec Roelke
 */
object CardImagePanel {
  /** Aspect ratio of a Magic: The Gathering card. */
  val AspectRatio = 63.0/88.0
  /** String to format for generating the link to a card's image on Scryfall. */
  val ScryfallFormat = "https://api.scryfall.com/cards/%s?format=image%s"
  /** String to format for generating the link to a card's image on Gatherer. */
  val GathererFormat = "https://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=%d&type=card%s"

  private val progressBars = collection.mutable.ArrayBuffer[JProgressBar]()
  private val progressLabels = collection.mutable.ArrayBuffer[JLabel]()

  /** @return the list of file names that will contain the image(s) to display for a card */
  def getFiles(card: Card): Seq[File] = SettingsDialog.settings.inventory.imageSource match {
    case "Scryfall" => (0 until card.imageNames.size).map((i) => Paths.get(SettingsDialog.settings.inventory.scans, s"${card(i).scryfallid}$i.jpg").toFile).toSeq
    case "Gatherer" => (0 until card.imageNames.size).map((i) => Paths.get(SettingsDialog.settings.inventory.scans, s"${card(i).multiverseid}$i.jpg").toFile).toSeq
    case _ => Seq.empty
  }

  /** @return the URLs to download a card's image(s) if they are missing */
  def getURLs(card: Card): Seq[Option[URL]] = SettingsDialog.settings.inventory.imageSource match {
    case "Scryfall" => card.layout match {
      case CardLayout.FLIP => Seq(Some(URL(ScryfallFormat.format(card.scryfallid, ""))))
      case CardLayout.MELD => (0 until card.imageNames.size).map((i) => Some(URL(ScryfallFormat.format(card(i).scryfallid, ""))))
      case _ => (0 until card.imageNames.size).map((i) => Some(URL(ScryfallFormat.format(card(i).scryfallid, if (i > 0 && i == card.imageNames.size - 1) "&face=back" else "")))).toSeq
    }
    case "Gatherer" => card.layout match {
      case CardLayout.FLIP => card.faces.map(_.multiverseid).zipWithIndex.map{ case (id, i) => if (id >= 0) {
        Some(URL(GathererFormat.format(id, if (i > 0 && i == card.faces.size - 1) "options=rotate180" else "")))
      } else {
        None
      }}
      case _ => card.faces.map((f) => Option.when(f.multiverseid >= 0)(URL(GathererFormat.format(f.multiverseid, ""))))
    }
    case _ => Seq.empty
  }

  /**
   * Create a progress bar and label indicating progress in downloading an image.
   * @return a panel containing the progress bar and image
   */
  def createStatusBar = {
    val panel = JPanel(FlowLayout(FlowLayout.LEFT, 0, 0))
    panel.setBorder(BorderFactory.createEtchedBorder())
    val bar = JProgressBar()
    bar.setEnabled(false)
    panel.add(bar)
    panel.add(Box.createHorizontalStrut(5))
    val label = JLabel()
    panel.add(label)

    progressBars += bar
    progressLabels += label
    panel
  }

  private case class DownloadRequest(source: CardImagePanel, card: Card)

  private object ImageDownloadWorker extends SwingWorker[Unit, Integer] {
    private val toDownload = java.util.concurrent.LinkedBlockingQueue[DownloadRequest]
    private var size = 0

    def downloadCard(source: CardImagePanel, card: Card) = {
      try {
        toDownload.put(DownloadRequest(source, card))
      } catch case e: InterruptedException => e.printStackTrace()
    }

    protected override def doInBackground() = {
      while (true) {
        val req = toDownload.take()
        val files = getFiles(req.card)
        val urls = getURLs(req.card)
        for (i <- 0 until urls.size) {
          if (!files(i).exists) {
            urls(i).foreach{ site =>
              files(i).getParentFile.mkdirs()
              try {
                val connection = site.openConnection()
                size = connection.getContentLength
                var downloaded = 0
                SwingUtilities.invokeLater(() => {
                  progressBars.foreach{ bar =>
                    bar.setEnabled(true)
                    bar.setMaximum(size)
                  }
                  progressLabels.foreach(_.setText(s"Downloading image of ${req.card.name} ..."))
                })

                Using.resources(BufferedInputStream(connection.getInputStream), BufferedOutputStream(FileOutputStream(files(i)))){ (in, out) =>
                  val data = new Array[Byte](1024)
                  var x = 0
                  while ({ x = in.read(data); x } > 0) {
                    out.write(data, 0, x)
                    downloaded += x
                    publish(downloaded)
                  }
                }
              } catch case e: Exception => System.err.println(s"Error downloading ${files(i)}: ${e.getMessage}")
            }
          }
        }
        if (req.card.layout == CardLayout.FLIP && files(0).exists) {
          try {
            val original = ImageIO.read(files(0))
            val flipped = BufferedImage(original.getWidth, original.getHeight, original.getType)
            val op = AffineTransformOp(AffineTransform.getRotateInstance(math.Pi, flipped.getWidth/2, flipped.getHeight/2), AffineTransformOp.TYPE_BILINEAR)
            ImageIO.write(op.filter(original, flipped), "jpg", files(1))
          } catch case e: Exception => e.printStackTrace()
        }
        if (SettingsDialog.settings.inventory.imageLimitEnable) {
          var images: Array[File] = null // Gets assigned in the very next statement, so null should be okay here
          while ({ images = Paths.get(SettingsDialog.settings.inventory.scans).toFile.listFiles; images.size > SettingsDialog.settings.inventory.imageLimit }) {
            images.minBy(_.lastModified).delete()
          }
        }
        SwingUtilities.invokeLater(() => {
          if (req.source.card.exists(_ == req.card))
            req.source.loadImages()
        })
        publish(0)
        if (toDownload.isEmpty) {
          SwingUtilities.invokeLater(() => {
            progressBars.foreach(_.setEnabled(false))
            progressLabels.foreach(_.setText(""))
          })
        }
      }
    }

    override protected def process(chunks: java.util.List[Integer]) = progressBars.foreach(_.setValue(chunks.asScala.last))

    execute()
  }
}

/**
 * Panel for displaying card images.  The card image will resize itself to fit the panel, but maintain its aspect ratio.
 * 
 * @constructor create a new card image panel, optionally displaying a card
 * @param card card to display, if defined
 * 
 * @author Alec Roelke
 */
class CardImagePanel(private var card: Option[Card] = None) extends JPanel {
  import CardImagePanel._

  private var image: Option[BufferedImage] = None
  private var faceImages = Seq.empty[Option[BufferedImage]]
  private var face = 0

  addMouseListener(new MouseAdapter {
    override def mousePressed(e: MouseEvent) = {
      face = card.map((face + 1) % _.imageNames.size).getOrElse(0)
      getParent.revalidate()
      repaint()
    }
  })
  card.foreach(setCard(_))

  /**
   * Change the card to display.  If the image has not been downloaded, download it first.
   * @param c card to display
   */
  def setCard(c: Card) = if (!card.exists(_ == c)) {
    card = Some(c)
    face = 0
    faceImages = Seq.empty

    try {
      Files.createDirectories(Path.of(SettingsDialog.settings.inventory.scans))
      if (getFiles(card.get).forall(_.exists))
        loadImages()
      else
        ImageDownloadWorker.downloadCard(this, c)
    } catch case e: IOException => {}
    revalidate()
    repaint()
  }

  /** Clear the panel so it displays nothing. */
  def clearCard() = {
    card = None
    face = 0
    faceImages = Seq.empty
    revalidate()
    repaint()
  }

  private def loadImages() = synchronized {
    card.foreach{ c =>
      faceImages = getFiles(c).map{ file => if (file.exists) {
        try {
          Some(ImageIO.read(file))
        } catch case e: IOException => None
      } else None }
      Option(getParent).foreach{ p => SwingUtilities.invokeLater(() => {
        p.revalidate()
        repaint()
      })}
    }
  }

  override def getPreferredSize = (Option(getParent), image) match {
    case (Some(p), Some(img)) => Dimension((p.getHeight*(img.getWidth.toDouble/img.getHeight.toDouble)).toInt, p.getHeight)
    case _ => super.getPreferredSize
  }

  override def setBounds(x: Int, y: Int, width: Int, height: Int) = {
    super.setBounds(x, y, width, height)

    if (!card.isDefined || width == 0 || height == 0)
      image = None
    else {
      val (w: Int, h: Int) = if (faceImages.size <= face || faceImages(face).isEmpty)
        ((height*AspectRatio).toInt, height)
      else
        faceImages(face).map((img) => (img.getWidth, img.getHeight)).getOrElse((0, 0))

      image = Some(BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB))
      val g = image.get.createGraphics
      if (faceImages.size <= face || !faceImages(face).isDefined) {
        val faceWidth = (h*AspectRatio).toInt

        val missingCardPane = JTextPane()
        val document = missingCardPane.getDocument.asInstanceOf[StyledDocument]
        val textStyle = document.addStyle("text", null)
        StyleConstants.setFontFamily(textStyle, UIManager.getFont("Label.font").getFamily)
        StyleConstants.setFontSize(textStyle, ComponentUtils.TextSize)
        val reminderStyle = document.addStyle("reminder", textStyle)
        StyleConstants.setItalic(reminderStyle, true)
        card.foreach(_.formatDocument(document, false, face))
        missingCardPane.setSize(Dimension(faceWidth - 4, h - 4))

        val img = BufferedImage(faceWidth, h, BufferedImage.TYPE_INT_ARGB)
        missingCardPane.paint(img.getGraphics)
        g.drawImage(img, 2, 2, null)
        g.setColor(Color.BLACK)
        g.drawRect(0, 0, faceWidth - 1, h - 1)
      }
      else
        faceImages(face).foreach(g.drawImage(_, 0, 0, null))
    }
  }

  override protected def paintComponent(g: Graphics) = {
    super.paintComponent(g)

    image.foreach{ img => g match {
      case g2: Graphics2D =>
        g2.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BICUBIC)
        val aspectRatio = img.getWidth.toDouble/img.getHeight.toDouble
        var width = (getHeight*aspectRatio).toInt
        var height = getHeight
        if (width > getWidth) {
          width = getWidth
          height = (width/aspectRatio).toInt
        }
        g2.drawImage(img, (getWidth - width)/2, (getHeight - height)/2, width, height, null)

        if (card.get.imageNames.size > 1) {
          val size = 15
          val border = 3
          FunctionalSymbol.values(if (face % 2 == 0) "T" else "Q").getIcon(size).paintIcon(this, g2, getWidth - size - border, getHeight - size - border)
        }
    }}
  }
}